README.md
script.aeon.tools


To go in Startup.xml:

	<onload condition="System.HasAddon(script.aeon.tools)">RunScript(script.aeon.tools,daemon=True)</onload>


You will need to set somewhere the following window property for daemon to choose what color you want (or 'none' for nothing).

	SetProperty(cpa_aeon_set,overall_color,home)
	or
	SetProperty(cpa_aeon_set,frequent_color,home)

	SetProperty(cfa_aeon_set,overall_color,home)
	or
	SetProperty(cfa_aeon_set,frequent_color,home)


I would set mine via Startup.xml as so:

	<onload condition="String.IsEmpty(Skin.String(aeon_background_fx))">Skin.SetString(aeon_background_fx,overall_color)</onload>
	<onload condition="String.IsEmpty(Skin.String(aeon_focus_fx))">Skin.SetString(aeon_focus_fx,overall_color)</onload>
	<onload>SetProperty(cfa_aeon_set,$INFO[Skin.String(aeon_background_fx)],home)</onload>
	<onload>SetProperty(cpa_aeon_set,$INFO[Skin.String(aeon_focus_fx)],home)</onload>
	<onload>SetProperty(cpa_aeon_fallback,,home)</onload>


You won't need to init the addon with 'firstrun' (as in mimix Startup.xml).

The daemon always processes current 7977 and current 7978, so you have color from each in aeon_tools_Colorcpa & aeon_tools_Colorcfa, plus the complimentary colors in aeon_tools_CColorcpa & aeon_tools_CColorcfa

Plus now old colors in aeon_tools_OldColorcpa & aeon_tools_OldColorcfa, plus the complimentary old colors in aeon_tools_OldCColorcpa & aeon_tools_OldCColorcfa

Have changed 'ImageUpdating' to 'aeon_tools_ImageUpdating'


Hopefully you can see what to change to work with aeon.


Basically display all fanarts or bg art via multiimage id=7977
And all poster or forefront art via multiimage id=7978

It still processes separate colors for current song poster/fan art.


And you can process any other art via say:

	<onfocus>RunScript(script.aeon.tools,info=overall_color,id='"IMAGE_TO_PROCESS"',prefix=UNIQUE_IDENTIFIER)</onfocus>
	<onfocus>RunScript(script.aeon.tools,info=frequent_color,id='"IMAGE_TO_PROCESS"',prefix=UNIQUE_IDENTIFIER)</onfocus>

	so then you will have Window(home).Property(UNIQUE_IDENTIFIER.ImageColor) & Window(home).Property(UNIQUE_IDENTIFIER.ImageCColor)

	&

	Window(home).Property(UNIQUE_IDENTIFIER.ImageOldColor) & Window(home).Property(UNIQUE_IDENTIFIER.ImageOldCColor)


The tool makes a cache of shrunk image files for faster getcolor calcs, these are 128x128 (128x72 made for fanart as AR is maintained, so this nicely equals 10x10 full screen fill if needed with no clipping [native display gui is 1280x720!!] /useless info).



Get stuck ask via gmail :-)



changelog.txt
v1.0.0
 - Initial Release


resources/lib/Utils.py
import xbmc
import xbmcaddon
import xbmcgui
import xbmcvfs
import xbmcplugin
import os, sys
import simplejson
import hashlib
import urllib
import random
import math
from PIL import Image, ImageOps, ImageEnhance, ImageDraw, ImageStat, ImageFilter
from decimal import *
from xml.dom.minidom import parse

AEONT_ADDON = xbmcaddon.Addon()
AEONT_ADDON_ID = AEONT_ADDON.getAddonInfo('id')
AEONT_ADDON_LANGUAGE = AEONT_ADDON.getLocalizedString
AEONT_ADDON_DATA_PATH = os.path.join(xbmc.translatePath("special://profile/addon_data/%s" % AEONT_ADDON_ID))
HOME = xbmcgui.Window(10000)


def Random_Color():
    return "ff" + "%06x" % random.randint(0, 0xFFFFFF)


def Complementary_Color(hex_color):
    """Returns complementary RGB color [should be format((255!]
    rgb = [hex_color[2:4], hex_color[4:6], hex_color[6:8]]
    comp = [format((325 - int(a, 16)), '02x') for a in rgb]
    return "FF" + "%s" % ''.join(comp)
    Example:
    >>>complementaryColor('FFFFFF')
    '000000'
    """
    rgb = [hex_color[2:4], hex_color[4:6], hex_color[6:8]]
    comp = ['%02X' % (255 - int(a, 16)) for a in rgb]
    """
    if (int(comp[0], 16) > 99 and int(comp[0], 16) < 150 and
        int(comp[1], 16) > 99 and int(comp[1], 16) < 150 and
        int(comp[2], 16) > 99 and int(comp[2], 16) < 150):
            return "FFc2836d"
    """
    return "FF" + "%s" % ''.join(comp)


def RemoveQuotes(label):
    if label.startswith("'") and label.endswith("'") and len(label) > 2:
        label = label[1:-1]
        if label.startswith('"') and label.endswith('"') and len(label) > 2:
            label = label[1:-1]
    return label


def Overall_Color(filterimage):
    md5 = hashlib.md5(filterimage).hexdigest()
    filename = md5 + ".png"
    targetfile = os.path.join(AEONT_ADDON_DATA_PATH, filename)
    cachedthumb = xbmc.getCacheThumbName(filterimage)
    xbmc_vid_cache_file = os.path.join("special://profile/Thumbnails/Video", cachedthumb[0], cachedthumb)
    xbmc_cache_file = os.path.join("special://profile/Thumbnails/", cachedthumb[0], cachedthumb[:-4] + ".jpg")
    if filterimage == "":
        return "", ""
    if not xbmcvfs.exists(targetfile):
        img = None
        for i in range(1, 4):
            try:
                if xbmcvfs.exists(xbmc_cache_file):
                    
                    img = Image.open(xbmc.translatePath(xbmc_cache_file))
                    break
                elif xbmcvfs.exists(xbmc_vid_cache_file):
                    
                    img = Image.open(xbmc.translatePath(xbmc_vid_cache_file))
                    break
                else:
                    filterimage = urllib.unquote(filterimage.replace("image://", "")).decode('utf8')
                    if filterimage.endswith("/"):
                        filterimage = filterimage[:-1]
                    
                    xbmcvfs.copy(filterimage, targetfile)
                    img = Image.open(targetfile)
                    break
            except:
                xbmc.sleep(300)
        if not img:
            return "", ""
        img.thumbnail((128, 128), Image.ANTIALIAS)
        img = img.convert('RGB')
        img.save(targetfile)
    else:
        img = Image.open(targetfile)
    imagecolor = Get_Overall_Color(img)
    return imagecolor


def Frequent_Color(filterimage):
    md5 = hashlib.md5(filterimage).hexdigest()
    filename = md5 + ".png"
    targetfile = os.path.join(AEONT_ADDON_DATA_PATH, filename)
    cachedthumb = xbmc.getCacheThumbName(filterimage)
    xbmc_vid_cache_file = os.path.join("special://profile/Thumbnails/Video", cachedthumb[0], cachedthumb)
    xbmc_cache_file = os.path.join("special://profile/Thumbnails/", cachedthumb[0], cachedthumb[:-4] + ".jpg")
    if filterimage == "":
        return "", ""
    if not xbmcvfs.exists(targetfile):
        img = None
        for i in range(1, 4):
            try:
                if xbmcvfs.exists(xbmc_cache_file):
                    
                    img = Image.open(xbmc.translatePath(xbmc_cache_file))
                    break
                elif xbmcvfs.exists(xbmc_vid_cache_file):
                    
                    img = Image.open(xbmc.translatePath(xbmc_vid_cache_file))
                    break
                else:
                    filterimage = urllib.unquote(filterimage.replace("image://", "")).decode('utf8')
                    if filterimage.endswith("/"):
                        filterimage = filterimage[:-1]
                    
                    xbmcvfs.copy(filterimage, targetfile)
                    img = Image.open(targetfile)
                    break
            except:
                xbmc.sleep(300)
        if not img:
            return "", ""
        img.thumbnail((128, 128), Image.ANTIALIAS)
        img = img.convert('RGB')
        img.save(targetfile)
    else:
        img = Image.open(targetfile)
    imagecolor = Get_Frequent_Color(img)
    return imagecolor


def Get_Overall_Color(img):
    colour_tuple = [None, None, None]
    for channel in range(3):

        # Get data for one channel at a time
        pixels = img.getdata(band=channel)

        values = []
        for pixel in pixels:
            values.append(pixel)

        colour_tuple[channel] = clamp(sum(values) / len(values))

    return 'ff%02x%02x%02x' % tuple(colour_tuple)


def Get_Frequent_Color(img):
    w, h = img.size
    pixels = img.getcolors(w * h)

    most_frequent_pixel = pixels[0]

    for count, colour in pixels:
        if count > most_frequent_pixel[0]:
            most_frequent_pixel = (count, colour)

    return 'ff%02x%02x%02x' % tuple(most_frequent_pixel[1])


def clamp(x): 
    return max(0, min(x, 255))


def log(txt):
    if isinstance(txt, str):
        txt = txt.decode("utf-8")
    message = u'%s: %s' % (AEONT_ADDON_ID, txt)
    xbmc.log(msg=message.encode("utf-8"), level=xbmc.LOGDEBUG)


def prettyprint(string):
    log(simplejson.dumps(string, sort_keys=True, indent=4, separators=(',', ': ')))


default.py
import sys
import os
import xbmc
import xbmcgui
import xbmcaddon

AEONT_ADDON = xbmcaddon.Addon()
AEONT_ADDON_VERSION = AEONT_ADDON.getAddonInfo('version')
AEONT_ADDON_LANGUAGE = AEONT_ADDON.getLocalizedString
AEONT_ADDON_PATH = AEONT_ADDON.getAddonInfo('path').decode("utf-8")
AEONT_ADDON_ID = AEONT_ADDON.getAddonInfo('id')
AEONT_ADDON_DATA_PATH = os.path.join(xbmc.translatePath("special://profile/addon_data/%s" % AEONT_ADDON_ID))
HOME = xbmcgui.Window(10000)
sys.path.append(xbmc.translatePath(os.path.join(AEONT_ADDON_PATH, 'resources', 'lib')))

from Utils import *


class Aeon_Tools_Main:

    def __init__(self):
        log("version %s started" % AEONT_ADDON_VERSION)
        self._init_vars()
        self._parse_argv()
        if not xbmcvfs.exists(AEONT_ADDON_DATA_PATH):
            # addon data path does not exist...create it
            xbmcvfs.mkdir(AEONT_ADDON_DATA_PATH)
        if self.infos:
            self._StartInfoActions()
        if self.control == "plugin":
            xbmcplugin.endOfDirectory(self.handle)
        while self.daemon and not xbmc.abortRequested:
            if not HOME.getProperty("cpa_aeon_set") == 'none':
                self.aeont_image_now_cpa = xbmc.getInfoLabel("Control.GetLabel(7978)")
                if self.aeont_image_now_cpa == '' and HOME.getProperty("cpa_aeon_fallback") != '':
                    self.aeont_image_now_cpa = HOME.getProperty("cpa_aeon_fallback")
                if self.aeont_image_now_cpa != self.aeont_image_prev_cpa:
                    try:
                        self.aeont_image_prev_cpa = self.aeont_image_now_cpa
                        HOME.setProperty("aeon_tools_OldColorcpa", HOME.getProperty("aeon_tools_Colorcpa"))
                        HOME.setProperty("aeon_tools_OldCColorcpa", HOME.getProperty("aeon_tools_CColorcpa"))
                        HOME.setProperty('aeon_tools_ImageUpdating', '0')
                        if HOME.getProperty("cpa_aeon_set") == 'overall_color':
                            imagecolor = Overall_Color(self.aeont_image_now_cpa)
                            HOME.setProperty("aeon_tools_Colorcpa", imagecolor)
                            HOME.setProperty("aeon_tools_CColorcpa", Complementary_Color(imagecolor))
                        elif HOME.getProperty("cpa_aeon_set") == 'frequent_color':
                            imagecolor = Frequent_Color(self.aeont_image_now_cpa)
                            HOME.setProperty("aeon_tools_Colorcpa", imagecolor)
                            HOME.setProperty("aeon_tools_CColorcpa", Complementary_Color(imagecolor))
                    except:
                        log("Could not process image for cpa daemon")
                    HOME.setProperty('aeon_tools_ImageUpdating', '1')
            if not HOME.getProperty("cfa_aeon_set") == 'none':
                self.aeont_image_now_cfa = xbmc.getInfoLabel("Control.GetLabel(7977)")
                if self.aeont_image_now_cfa != self.aeont_image_prev_cfa:
                    try:
                        self.aeont_image_prev_cfa = self.aeont_image_now_cfa
                        HOME.setProperty("aeon_tools_OldColorcfa", HOME.getProperty("aeon_tools_Colorcfa"))
                        HOME.setProperty("aeon_tools_OldCColorcfa", HOME.getProperty("aeon_tools_CColorcfa"))
                        HOME.setProperty('aeon_tools_ImageUpdating', '0')
                        if HOME.getProperty("cfa_aeon_set") == 'overall_color':
                            imagecolor = Overall_Color(self.aeont_image_now_cfa)
                            HOME.setProperty("aeon_tools_Colorcfa", imagecolor)
                            HOME.setProperty("aeon_tools_CColorcfa", Complementary_Color(imagecolor))
                        elif HOME.getProperty("cfa_aeon_set") == 'frequent_color':
                            imagecolor = Frequent_Color(self.aeont_image_now_cfa)
                            HOME.setProperty("aeon_tools_Colorcfa", imagecolor)
                            HOME.setProperty("aeon_tools_CColorcfa", Complementary_Color(imagecolor))
                    except:
                        log("Could not process image for cfa daemon")
                    HOME.setProperty('aeon_tools_ImageUpdating', '1')
            self.aeont_image_now = xbmc.getInfoLabel("Player.Art(thumb)")
            self.aeont_image_now_fa = xbmc.getInfoLabel("MusicPlayer.Property(Fanart_Image)")
            if self.aeont_image_now != self.aeont_image_prev and xbmc.Player().isPlayingAudio():
                try:
                    self.aeont_image_prev = self.aeont_image_now
                    imagecolor = Overall_Color(self.aeont_image_now)
                    HOME.setProperty("ImageColor1", imagecolor)
                except:
                    log("Could not process image for f daemon")
            if self.aeont_image_now_fa != self.aeont_image_prev_fa and xbmc.Player().isPlayingAudio():
                try:
                    self.aeont_image_prev_fa = self.aeont_image_now_fa
                    imagecolor = Overall_Color(self.aeont_image_now_fa)
                    HOME.setProperty("ImageColorfa1", imagecolor)
                except:
                    log("Could not process image for fa daemon")
            xbmc.sleep(300)

    def _StartInfoActions(self):
        HOME.setProperty(self.prefix + 'aeon_tools_ImageUpdating', '0')
        for info in self.infos:
            if info == 'overall_color':
                HOME.setProperty(self.prefix + "ImageOldColor", HOME.getProperty(self.prefix + "ImageOldColor"))
                HOME.setProperty(self.prefix + "ImageOldCColor", HOME.getProperty(self.prefix + "ImageOldCColor"))
                imagecolor = Overall_Color(self.id)
                HOME.setProperty(self.prefix + "ImageColor", imagecolor)
                HOME.setProperty(self.prefix + "ImageCColor", Complementary_Color(imagecolor))
            elif info == 'frequent_color':
                HOME.setProperty(self.prefix + "ImageOldColor", HOME.getProperty(self.prefix + "ImageOldColor"))
                HOME.setProperty(self.prefix + "ImageOldCColor", HOME.getProperty(self.prefix + "ImageOldCColor"))
                imagecolor = Frequent_Color(self.id)
                HOME.setProperty(self.prefix + "ImageColor", imagecolor)
                HOME.setProperty(self.prefix + "ImageCColor", Complementary_Color(imagecolor))
        HOME.setProperty(self.prefix + 'aeon_tools_ImageUpdating', '1')

    def _init_vars(self):
        self.window = xbmcgui.Window(10000)  # Home Window
        self.control = None
        self.infos = []
        self.id = ""
        self.dbid = ""
        self.prefix = ""
        self.daemon = False
        self.aeont_image_now = ""
        self.aeont_image_now_fa = ""
        self.aeont_image_now_cfa = ""
        self.aeont_image_now_cpa = ""
        self.aeont_image_prev = ""
        self.aeont_image_prev_fa = ""
        self.aeont_image_prev_cfa = ""
        self.aeont_image_prev_cpa = ""
        self.autoclose = ""

    def _parse_argv(self):
        args = sys.argv
        for arg in args:
            arg = arg.replace("'\"", "").replace("\"'", "")
            if arg == 'script.aeon.tools':
                continue
            elif arg.startswith('daemon='):
                self.daemon = True
            elif arg.startswith('prefix='):
                self.prefix = arg[7:]
                if not self.prefix.endswith("."):
                    self.prefix = self.prefix + "."
            elif arg.startswith('container='):
                self.container = RemoveQuotes(arg[10:])

class Aeon_Tools_Monitor(xbmc.Monitor):

    def __init__(self, *args, **kwargs):
        xbmc.Monitor.__init__(self)

    def onPlayBackStarted(self):
        pass


if __name__ == "__main__":
    Aeon_Tools_Main()
log('finished')
